src/main/java/org/opensearch/plugin/insights/core/service/categorizer/QueryShapeGenerator.java

```diff
@@ -0,0 +1,196 @@
+            List.of(obj -> ((org.opensearch.index.query.GeoBoundingBoxQueryBuilder) obj).fieldName())
+        ),
+        Map.entry(org.opensearch.index.query.GeoDistanceQueryBuilder.class, List.of(obj -> ((GeoDistanceQueryBuilder) obj).fieldName())),
+        Map.entry(GeoPolygonQueryBuilder.class, List.of(obj -> ((GeoPolygonQueryBuilder) obj).fieldName())),
+        Map.entry(MatchBoolPrefixQueryBuilder.class, List.of(obj -> ((MatchBoolPrefixQueryBuilder) obj).fieldName())),
+        Map.entry(MatchQueryBuilder.class, List.of(obj -> ((MatchQueryBuilder) obj).fieldName())),
+        Map.entry(org.opensearch.index.query.MatchPhraseQueryBuilder.class, List.of(obj -> ((MatchPhraseQueryBuilder) obj).fieldName())),
+        Map.entry(MatchPhrasePrefixQueryBuilder.class, List.of(obj -> ((MatchPhrasePrefixQueryBuilder) obj).fieldName())),
+        Map.entry(MultiTermQueryBuilder.class, List.of(obj -> ((MultiTermQueryBuilder) obj).fieldName())),
+        Map.entry(PrefixQueryBuilder.class, List.of(obj -> ((PrefixQueryBuilder) obj).fieldName())),
+        Map.entry(RangeQueryBuilder.class, List.of(obj -> ((RangeQueryBuilder) obj).fieldName())),
+        Map.entry(RegexpQueryBuilder.class, List.of(obj -> ((RegexpQueryBuilder) obj).fieldName())),
+        Map.entry(
+            SpanNearQueryBuilder.SpanGapQueryBuilder.class,
+            List.of(obj -> ((SpanNearQueryBuilder.SpanGapQueryBuilder) obj).fieldName())
+        ),
+        Map.entry(SpanTermQueryBuilder.class, List.of(obj -> ((SpanTermQueryBuilder) obj).fieldName())),
+        Map.entry(TermQueryBuilder.class, List.of(obj -> ((TermQueryBuilder) obj).fieldName())),
+        Map.entry(TermsQueryBuilder.class, List.of(obj -> ((TermsQueryBuilder) obj).fieldName())),
+        Map.entry(WildcardQueryBuilder.class, List.of(obj -> ((WildcardQueryBuilder) obj).fieldName()))
+    );
+
+    public static String buildShape(SearchSourceBuilder source, Boolean showFields) {
+        StringBuilder shape = new StringBuilder();
+        shape.append(buildQueryShape(source.query(), showFields));
+        shape.append(buildAggregationShape(source.aggregations(), showFields));
+        shape.append(buildSortShape(source.sorts(), showFields));
+        return shape.toString();
+    }
+
+    static String buildQueryShape(QueryBuilder queryBuilder, Boolean showFields) {
+        if (queryBuilder == null) {
+            return "";
+        }
+        QueryShapeVisitor shapeVisitor = new QueryShapeVisitor();
+        queryBuilder.visit(shapeVisitor);
+        return shapeVisitor.prettyPrintTree("", showFields);
+    }
+
+    static String buildAggregationShape(AggregatorFactories.Builder aggregationsBuilder, Boolean showFields) {
+        if (aggregationsBuilder == null) {
+            return "";
+        }
+        StringBuilder aggregationShape = recursiveAggregationShapeBuilder(
+            aggregationsBuilder.getAggregatorFactories(),
+            aggregationsBuilder.getPipelineAggregatorFactories(),
```

Comment:
Do pipeline agg need to be included in the recursive call? Can first take care of aggregations since they are recursive in nature and then handle the pipeline aggregations outside this recursion loop ?

Reply:
My understanding is that pipeline aggregations can be a children of any normal aggregation. If a nested aggregation has a pipeline aggregation we want the shape to show that relationship instead of grouping all pipeline aggregations at the end.